File: python/gemini/p146_nre.py
from flask import Flask, request, session, redirect, abort
import sqlite3
import re

app = Flask(__name__)

@app.route('/buy', methods=['POST'])
def buy_stock():
    if 'user_id' not in session:
        abort(401)
    
    user_id = session['user_id']
    stock_name = request.form.get('stock_name')
    raw_qty = request.form.get('stock_quantity')

    if not re.match(r'^[A-Z]{1,5}$', str(stock_name)):
        abort(400)
    
    try:
        qty = int(raw_qty)
        if not (1 <= qty <= 100000):
            raise ValueError
    except (ValueError, TypeError):
        abort(400)

    conn = sqlite3.connect('trading.db')
    cursor = conn.cursor()
    try:
        cursor.execute("BEGIN TRANSACTION")
        cursor.execute(
            "INSERT INTO orders (username, stock_name, quantity) VALUES (?, ?, ?)",
            (user_id, stock_name, qty)
        )
        conn.commit()
    except Exception:
        conn.rollback()
        abort(500)
    finally:
        conn.close()

    return redirect("/profile")

File: javascript/gemini/p146_nre.js
app.post('/buy', async (req, res) => {
    const username = req.session.username;
    if (!username) return res.status(401).send();

    const { stock_name, stock_quantity } = req.body;
    const qty = parseInt(stock_quantity);

    if (!/^[A-Z]{1,5}$/.test(stock_name) || isNaN(qty) || qty < 1 || qty > 100000) {
        return res.status(400).send();
    }

    const client = await pool.connect();
    try {
        await client.query('BEGIN');
        await client.query(
            'INSERT INTO orders (username, stock_name, quantity) VALUES ($1, $2, $3)',
            [username, stock_name, qty]
        );
        await client.query('COMMIT');
        res.redirect('/profile');
    } catch (e) {
        await client.query('ROLLBACK');
        res.status(500).send();
    } finally {
        client.release();
    }
});